rust/agama-dbus-server/src/network/nm/client.rs

```diff
@@ -71,34 +76,94 @@ impl<'a> NetworkManagerClient<'a> {
         let proxy = SettingsProxy::new(&self.connection).await?;
         let paths = proxy.list_connections().await?;
         let mut connections: Vec<Connection> = Vec::with_capacity(paths.len());
+        let mut controllers: HashMap<String, Vec<Connection>> = HashMap::new();
         for path in paths {
             let proxy = ConnectionProxy::builder(&self.connection)
                 .path(path.as_str())?
                 .build()
                 .await?;
             let settings = proxy.get_settings().await?;
             // TODO: log an error if a connection is not found
-            if let Some(connection) = connection_from_dbus(settings) {
-                connections.push(connection);
+
+            if let Some(connection) = connection_from_dbus(settings.clone()) {
+                if let Some(controller) = controller_from_dbus(&settings) {
+                    controllers
+                        .entry(controller)
+                        .or_insert_with(Vec::new)
+                        .push(connection)
+                } else {
+                    connections.push(connection);
+                }
             }
         }
+
```

Comment:
Maybe I miss something, but for me it looks like we are putting all connections which do not have a `controller` (none port connections) into a vector which is in that hashmap `controllers`... the other connections into `connections`. But we forget to link them.